src/overlay/SurveyDataManager.cpp

```diff
@@ -0,0 +1,484 @@
+    , mInitialUniqueFetchBytesRecv(peerMetrics.mUniqueFetchBytesRecv)
+    , mInitialDuplicateFetchBytesRecv(peerMetrics.mDuplicateFetchBytesRecv)
+    , mInitialUniqueFloodMessageRecv(peerMetrics.mUniqueFloodMessageRecv)
+    , mInitialDuplicateFloodMessageRecv(peerMetrics.mDuplicateFloodMessageRecv)
+    , mInitialUniqueFetchMessageRecv(peerMetrics.mUniqueFetchMessageRecv)
+    , mInitialDuplicateFetchMessageRecv(peerMetrics.mDuplicateFetchMessageRecv)
+{
+}
+
+SurveyDataManager::SurveyDataManager(VirtualClock const& clock,
+                                     medida::Meter const& lostSyncMeter)
+    : mClock(clock), mLostSyncMeter(lostSyncMeter)
+{
+}
+
+bool
+SurveyDataManager::startSurveyRecording(
+    TimeSlicedSurveyStartRecordingMessage const& msg,
+    std::map<NodeID, Peer::pointer> const& peers)
+{
+    ZoneScoped;
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+
+    if (phase() == SurveyPhase::INACTIVE)
+    {
+        CLOG_INFO(Overlay, "Starting survey recording with nonce {}",
+                  msg.nonce);
+        mRecordStartTime = mClock.now();
+        mNonce = msg.nonce;
+        mSurveyor = msg.surveyorID;
+        mRecordingNodeData.emplace(mLostSyncMeter.count());
+        if (mRecordingInboundPeerData.empty() &&
+            mRecordingOutboundPeerData.empty())
+        {
+            for (auto const& [_, peer] : peers)
+            {
+                // Copy initial peer metrics
+                getPeerDataMapFromPeer(*peer).try_emplace(
+                    peer->getPeerID(), peer->getPeerMetrics());
+            }
+            return true;
+        }
+
+        emitInconsistencyError("startSurveyRecording");
+        return false;
+    }
+
+    CLOG_INFO(Overlay,
+              "Ignoring request to start survey recording with nonce {} "
+              "because there is already an active survey",
+              msg.nonce);
+    return false;
+}
+
+bool
+SurveyDataManager::stopSurveyRecording(
+    TimeSlicedSurveyStopRecordingMessage const& msg,
+    std::map<NodeID, Peer::pointer> const& peers, Config const& config)
+{
+    ZoneScoped;
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+
+    uint32_t const nonce = msg.nonce;
+    if (phase() == SurveyPhase::RECORDING && mNonce == nonce &&
+        mSurveyor == msg.surveyorID)
+    {
+        CLOG_INFO(Overlay, "Stopping survey recording with nonce {}", nonce);
+        mRecordEndTime = mClock.now();
+
+        if (!mFinalInboundPeerData.empty() || !mFinalOutboundPeerData.empty())
+        {
+            emitInconsistencyError("stopSurveyRecording");
+            return false;
+        }
+
+        // Finalize peer data
+        for (auto const& [_, peer] : peers)
+        {
+            finalizePeerData(*peer);
+        }
+
+        finalizeNodeData(config);
+
+        // Clear recording data
+        mRecordingInboundPeerData.clear();
+        mRecordingOutboundPeerData.clear();
+
+        return true;
+    }
+    CLOG_INFO(Overlay,
```

Comment:
I'm not sure. It's a condition that anyone trusted to run a survey could trigger. It's harmless because it's easy and cheap to defend against, but turning this into a warning might make someone thing something is wrong with their node when it's another node that's misbehaving. In general, does stellar-core emit warnings when peers misbehave? If so, this should probably be a warning.

Reply:
Done in ef5282cf854e9a6e5a2e55849571c01f109d7390